Simplified/NYPLAccountSignInViewController.m

```diff
@@ -215,24 +221,80 @@ - (void)viewDidLoad
+    }
+  };
+
+  NSMutableArray *section0AcctInfo;
+  if (!self.businessLogic.selectedAuthentication.needsAuth && self.businessLogic.selectedAuthentication) {
+    section0AcctInfo = @[].mutableCopy;
+  } else if (self.businessLogic.selectedAuthentication && self.businessLogic.isSignedIn) {
+    // user already logged in
+    // show only the selected auth method
+    section0AcctInfo = @[].mutableCopy;
+    insertCredentials(self.businessLogic.selectedAuthentication, section0AcctInfo);
+  } else if (!self.businessLogic.isSignedIn && self.businessLogic.userAccount.needsAuth) {
+    // user needs to sign in
+    section0AcctInfo = @[].mutableCopy;
+
+    NSUInteger samlIndex = [self.businessLogic.libraryAccount.details.auths indexOfObjectPassingTest:^BOOL(AccountDetailsAuthentication * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
+      return obj.samlIdps.count > 0;
+    }];
+
+    if (samlIndex != NSNotFound) {
+      NSString *libraryInfo = [NSString stringWithFormat:@"Log in to %@ required to download materials.", self.businessLogic.libraryAccount.name];
+      [section0AcctInfo addObject:[[InfoHeaderCellType alloc] initWithInformation:libraryInfo]];
+    }
+
+    if (self.businessLogic.libraryAccount.details.auths.count > 1) {
+      // multiple authentication methods
+      for (AccountDetailsAuthentication *authenticationMethod in self.businessLogic.libraryAccount.details.auths) {
+        // show all possible login methods
+        AuthMethodCellType *autheticationCell = [[AuthMethodCellType alloc] initWithAuthenticationMethod:authenticationMethod];
+        [section0AcctInfo addObject:autheticationCell];
+        if (authenticationMethod.methodDescription == self.businessLogic.selectedAuthentication.methodDescription) {
+          // selected method, unfold
+          insertCredentials(authenticationMethod, section0AcctInfo);
+        }
+      }
+    } else if (self.businessLogic.libraryAccount.details.auths.count == 1) {
+      // only 1 authentication method
+      // no method header needed
+      insertCredentials(self.businessLogic.libraryAccount.details.auths[0], section0AcctInfo);
+    } else if (self.businessLogic.selectedAuthentication) {
+      // only 1 authentication method
+      // no method header needed
+      insertCredentials(self.businessLogic.selectedAuthentication, section0AcctInfo);
+    }
   } else {
-    //Server expects a blank string. Passes local textfield validation.
-    self.PINTextField.text = @"";
-    section0 = @[@(CellKindBarcode),
-                 @(CellKindLogInSignOut)];
+    section0AcctInfo = @[].mutableCopy;
+    insertCredentials(self.businessLogic.selectedAuthentication, section0AcctInfo);
   }
+
   NSArray *section1;
   if ([self.businessLogic registrationIsPossible]) {
     section1 = @[@(CellKindRegistration)];
   } else {
     section1 = @[];
   }
-  self.tableData = @[section0, section1];
+  self.tableData = @[section0AcctInfo, section1];
+  [self.tableView reloadData];
```

Comment:
yes, it does get called multiple times in case of changes to user account (setupTableData is part of `accountDidChange` observer). I didn't notice any issues related to that though. And given that `setupTableData` intention is to generate a new table definition, I'd rather refresh the table too much, than risk, for example, an _index out of bound_ error in cellForRow when accessing `tableData`

Reply:
sounds good.